pkg/cdc/generate_config.go

```diff
@@ -0,0 +1,143 @@
+package cdc
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
+)
+
+// UpdateConfig updates the json formatted Cassandra config which incorporates the JVM options (under key additional-jvm-opts) passed into the launch scripts
+// for Cassandra via the config builder.
+func UpdateConfig(config json.RawMessage, cassDC cassdcapi.CassandraDatacenter) (json.RawMessage, error) {
```

Comment:
I can't follow the logic of doing it this way. What was the main reason to not do these in the existing methods such as GetConfigASJson() ?

Reply:
I don't quite understand the question here. `GetConfigAsJson()` looks like it would be a very appropriate place to hook this logic in, but we wouldn't want to put all of the reconciliation logic in that function would we? It seems nicer to have it segregated off in `pkg`.

Can you clarify what you're suggesting we change?

Reply:
Just creating the JSON config. Why would it do reconcile logic otherwise? It's just taking whatever has been set in the CassandraDatacenter and creating a JSON output of configuration - it does not care about any logic otherwise. 

It would reduce quite a lot of the logic, since there's no need to read any JSON - everything is still in a struct form there.

Reply:
I'm still not understanding, sorry. When the CDC config changes, reconciliation of the CONFIG_DATA environment variable is required to bring it into the state declared in the CR.

The JSON still needs to be read because (in the case CDC is disabled) the CDC related configs (in cassandra-env.sh) need to be removed.

I don't understand what you're looking to change here?

Reply:
Why would you need to remove them? If there's no CDC definition set -> ConfigASJSON does not generate anything to the JSON about CDC. If it is set, then it is generated there. If the newly generated value does not match the existing one -> StatefulSet is reconciled and modified to get the new config.

In which case would this not be true? Why would ConfigASJSON generate a config with something that's no longer set and it needs to be removed?

Reply:
I've been considering this one. The problem is that the function `serverconfig.GetModelValues` which generates the JSON is not very well structured. 

It is in a strange place, has far too many arguments, and produces an untyped `type NodeConfig map[string]interface{}`.

I followed this path for a little while regardless. I thought I could add an argument `CDCConfig cassdcapi.CDCConfiguration` to `GetModelValues` and generate the json from there. However, when I do this, I get a circular import whereby `serverconfig` imports `cassdcapi` and vice-versa.

The root of the problem is that the API package `apis/cassandra/v1beta1` is polluted with excessive numbers of functions (especially `ConfigASJSON`, which should probably be segregated from the APIs). This makes it hard to use `ConfigASJSON` to assemble the CONFIG_DATA env variable because there is no clean way to pass in all of the required fields (in the pulsar case there is a substantial number). You can never use the CR-linked struct itself if you want to call anything to do with `serverconfig.GetModelValues`.

The fix for this is to move `ConfigASJSON` into its own package, and address the questionable number of parameters and loosely typed return value of `serverconfig.GetModelValues`. I think we should actually build some sort of pipeline construct, whereby each phase performs some transformation on the config and passes it to the next.

All of this is major refactoring and I don't think it belongs in this ticket.